### PR TITLE
Tab completion NFA state graph visualization tool.

### PR DESCRIPTION
**Pull request by georgebrock**
The tab completion system that was introduced in 1b673e5 uses a
Non-deterministic Finite Automaton (NFA) to understand the user's input and
select relevant completions. The NFA has a directed graph of states.

While the current NFA's state graph is small, the intention is to replace it
with a much larger and more complex graph.

This commit introduces a visualization tool, which outputs a representation
of the NFA's state graph in Graphviz's dot language.

---

Current output:

Tab completion graph

### CODE REVIEW

File: lib/gitsh/tab_completion/README.md [hunk 1]
```diff
@@ -102,6 +102,21 @@ then hit <kbd>tab</kbd>. In this case, we'd pass `add` to the automaton and move
 from state `0` to state `4`. From state `4` the next thing we expect is a path,
 so we'd offer file paths from the current directory as completions.
 
+### Building the state graph
+
+The state graph is built by the `Gitsh::TabCompletion::AutomatonFactory` class.
+The initial implementation is small and fixed, but is expected to grow in size
+and complexity as we support more subtle forms of tab completion.
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/README.md]: I'm not excited about this sentence/I worry about it staying around longer than this graph's simplicity.

File: lib/gitsh/tab_completion/README.md [hunk 2]
```diff
@@ -102,6 +102,21 @@ then hit <kbd>tab</kbd>. In this case, we'd pass `add` to the automaton and move
...
+
+To help debug the state graph, the repository includes a visualisation tool.
+Running `bin/tcviz` will load the configuration files, and output the graph in
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/README.md]: No comma.

File: lib/gitsh/tab_completion/visualization.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      attr_reader :start_state
+
+      def node_labels
+        @node_labels ||= states.map do |state|
+          '%s [ label=%s ];' % [state_name(state), state.name.inspect]
+        end
+      end
+
+      def transitions
+        @transitions ||= states.inject([]) do |acc, state|
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/visualization.rb]: This is a `#flat_map` right?

**Comment by georgebrock** [comment]
> I kinda expected to find `#to_dot` methods on `State`, matchers, `Automaton`, etc.

You make a good point. I originally wrote this as a throw-away script (there were a lot of `#send` calls, and it even used `ObjectSpace` to find all the `State` instances). It was surprisingly useful in debugging the `nfa-dsl-omg` branch, so I decided to keep it around, but didn't refactor as far as I should have.

I'll take another pass at it.

File: lib/gitsh/tab_completion/README.md (hunk 2 above)
Comment (georgebrock) [inline lib/gitsh/tab_completion/README.md]: Good point. Also, it doesn't load any configuration files (yet—that's on the larger branch I extracted this smaller piece from).

File: lib/gitsh/tab_completion/README.md (hunk 1 above)
Comment (georgebrock) [inline lib/gitsh/tab_completion/README.md]: I'll drop it.

**Comment by georgebrock** [comment]
@mike-burns I decided I still wanted to keep all of the dot formatting stuff in the same object, especially because it's only useful for debugging. To avoid exposing the internals of the automaton and its states, I used the visitor pattern. Maybe this is better?

File: lib/gitsh/tab_completion/visualization.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        def visit_free_transition(start_state, end_state)
+          transitions << '%s -> %s;' % [
+            state_name(start_state),
+            state_name(end_state),
+          ]
+        end
+
+        private
+
+        def state_name(state)
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/visualization.rb]: Maybe `state_key` or `state_label` or `state_identifier` are more appropriate names here.

Comment (georgebrock) [inline lib/gitsh/tab_completion/visualization.rb]: I like `state_identifier` (`label` has meaning in the dot language)

**Pull request merged**